File: algorithm/103.binary-tree-zigzag-level-order-traversal/103.binary-tree-zigzag-level-order-traversal.java
/*
 * @lc app=leetcode id=103 lang=java
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) {
            return res;            
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int level = 1;
        while(!queue.isEmpty()) {
            int n = queue.size();
            Deque<Integer> cur = new LinkedList<>();
            for(int i = 0; i < n; i++) {
                TreeNode node = queue.poll();
                if(level % 2 == 0) {
                    cur.offerFirst(node.val);
                } else {
                    cur.offerLast(node.val);
                }
                if(node.left != null) {
                    queue.offer(node.left);
                }
                if(node.right != null) {
                    queue.offer(node.right);
                }
            }
            if(cur.size() > 0) {
                res.add(new ArrayList<Integer>(cur));
            }
            level++;
        }
        return res;
    }
}
// @lc code=end

File: algorithm/103.binary-tree-zigzag-level-order-traversal/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        ans = []
        queue = deque() # queue
        queue.append(root)
        zigzag = True
        while queue:
            tmp = []
            # print(queue)
            for i in range(0, len(queue)):
                node = queue.popleft()
                tmp.append(node.val)
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
            ans.append(tmp if zigzag else tmp[::-1])
            zigzag = not zigzag
        return ans



# @lc code=end

